docs/pages/pmd/rules/java/performance.md

```diff
@@ -113,6 +161,152 @@ or AdditiveExpression) and count(.//PrimaryPrefix/Name)=1]
+```
+
+## AvoidConcatInAppend
+
+**Since:** PMD 6.17.0
+
+**Priority:** Medium High (2)
+
+Concatenation of Strings is used inside an StringBuilder.append argument. Problem: Each statement with one or more +-operators creates a hidden temporary StringBuilder, a char[] and a new String object, which all have to be garbage collected.
+            Solution: Use an extra fluent append instead of concatenation.
+
+**This rule is defined by the following XPath expression:**
+``` xpath
+//MethodDeclaration/Block/BlockStatement[
+(.//PrimaryExpression/PrimaryPrefix/Name[substring-after(@Image, '.') = 'append']
+or
+.//PrimaryExpression/PrimarySuffix/@Image = 'append')
+and
+.//PrimaryExpression/PrimarySuffix/Arguments/ArgumentList/Expression[1]/AdditiveExpression[@Image='+'
+and (count(PrimaryExpression/PrimaryPrefix/Name) > 0)
+and not(PrimaryExpression/PrimaryPrefix/Name/@Image=
+ancestor::ClassOrInterfaceBody//FieldDeclaration[@Final='true']//VariableDeclaratorId/@Image)
+and not(PrimaryExpression/PrimaryPrefix/Name/@Image=
+ancestor::Block//LocalVariableDeclaration[@Final='true']//VariableDeclaratorId/@Image)
+]]
+```
+
+**Example(s):**
+
+``` java
+{%raw%}public class StringStuff {
+   private String bad(String arg) {
+        StringBuilder sb = new StringBuilder();
+        sb.append("arg='" + arg + "'");
+        return sb.toString();
+   }
+
+   private String good(String arg) {
+        StringBuilder sb = new StringBuilder();
+        sb.append("arg='").append(arg).append("'");
+        return sb.toString();
+   }
+}{%endraw%}
+```
+
+**Use this rule by referencing it:**
+``` xml
+<rule ref="category/java/performance.xml/AvoidConcatInAppend" />
+```
+
+## AvoidConcatInLoop
+
+**Since:** PMD 6.17.0
+
+**Priority:** Medium High (2)
+
+A String is built in a loop by concatenation. Problem: Each statement with one or more +-operators creates a hidden temporary StringBuilder, a char[] and a new String object, which all have to be garbage collected. 
+            Solution: Use the StringBuilder append method.
+
+**This rule is defined by the following XPath expression:**
+``` xpath
+(//ForStatement | //WhileStatement | //DoStatement)//AssignmentOperator[
+    (: a += ...;  -- a being a string :)
+    @Image='+=' and preceding-sibling::*[1]/PrimaryPrefix/Name[pmd-java:typeIs('java.lang.String')]
+
+    (: a = ... + a + ...; -- a being a string :)
+    or @Image='=' and following-sibling::*[1]/AdditiveExpression/PrimaryExpression/PrimaryPrefix/Name[
+        pmd-java:typeIs('java.lang.String')
+        and @Image = ancestor::StatementExpression/PrimaryExpression/PrimaryPrefix/Name/@Image
+    ]
+]/.. (: Go up to report on the StatementExpression :)
+```
+
+**Example(s):**
+
+``` java
+{%raw%}public class StringStuff {
+   private String bad(String arg) {
+        String log = "";
+        List<String> values = Arrays.asList("tic ", "tac ", "toe ");
+        for (String val : values) {
+            log += val;
```

Comment:
Is string concatenation is still an issue nowadays (2019) ? I've heard that there were optimizations for that in JDK9, for example.

Reply:
There is indeed some optimization done with Java 9:
* http://openjdk.java.net/jeps/280
* https://stackoverflow.com/questions/46512888/how-is-string-concatenation-implemented-in-java-9

Before Java9, inside the loop, the code generated was basically:
```
StringBuilder temp = new StringBuilder();
log = temp.append(log).append(val).toString();
```

Since Java9, it looks more like (roughly):
```
log = StringConcatFactory.makeConcatWithConstants(..., log, val);
```

So, the rule is definitely less relevant for java9+.

Reply:
it's worth noting, that this specific example would be unaffected by Java 9+ optimizations, as the concat is done within a loop